File: CLIENT/SPRINGBOOT PLUGIN/src/main/java/com/alexistdev/geolicense/starter/service/LicenseVerificationService.java
package com.alexistdev.geolicense.starter.service;

import com.alexistdev.geolicense.starter.properties.GeoLicenseProperties;
import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.time.Duration;

public class LicenseVerificationService {

    private static final Logger log = LoggerFactory.getLogger(LicenseVerificationService.class);

    private final GeoLicenseProperties properties;
    private final LicenseHolder licenseHolder;
    private final MachineIdGenerator machineIdGenerator;
    private final ObjectMapper objectMapper = new ObjectMapper();
    private final HttpClient httpClient = HttpClient.newBuilder()
            .connectTimeout(Duration.ofSeconds(10))
            .build();

    public LicenseVerificationService(GeoLicenseProperties properties,
                                      LicenseHolder licenseHolder,
                                      MachineIdGenerator machineIdGenerator) {
        this.properties = properties;
        this.licenseHolder = licenseHolder;
        this.machineIdGenerator = machineIdGenerator;
    }

    public void verify() {
        String token = licenseHolder.getToken();
        if (token == null) {
            log.warn("Verification skipped — no license token present");
            licenseHolder.setValid(false);
            return;
        }

        String machineId = machineIdGenerator.generate();
        String body = objectMapper.createObjectNode()
                .put("token", token)
                .put("machineId", machineId)
                .toString();

        try {
            HttpRequest request = HttpRequest.newBuilder()
                    .uri(URI.create(properties.getServerUrl() + "/api/v1/licenses/verify"))
                    .header("Content-Type", "application/json")
                    .POST(HttpRequest.BodyPublishers.ofString(body))
                    .timeout(Duration.ofSeconds(15))
                    .build();

            HttpResponse<String> response = httpClient.send(request, HttpResponse.BodyHandlers.ofString());
            JsonNode root = objectMapper.readTree(response.body());
            boolean status = root.path("status").asBoolean(false);

            if (response.statusCode() == 200 && status) {
                licenseHolder.setValid(true);
                log.debug("License verified successfully");
            } else {
                licenseHolder.setValid(false);
                String message = root.path("messages").path(0).asText("unknown reason");
                log.warn("License verification failed: {}", message);
            }

        } catch (Exception e) {
            log.warn("License server unreachable during verification: {}", e.getMessage());
        }
    }
}

File: CLIENT/LARAVEL/app/Services/GeoLicense/LicenseVerificationService.php
<?php

namespace App\Services\GeoLicense;

use Illuminate\Support\Facades\Http;
use Illuminate\Support\Facades\Log;

class LicenseVerificationService
{
    public function __construct(
        private readonly LicenseHolder      $holder,
        private readonly MachineIdGenerator $machineId,
    ) {}

    /**
     * Calls POST /api/v1/licenses/verify on the GeoLicense server.
     * Updates the cache validity flag based on the response.
     *
     * If the server is unreachable, validity is left unchanged so the
     * grace period in LicenseValidationMiddleware can absorb the outage.
     */
    public function verify(): void
    {
        $token = $this->holder->getToken();

        if (empty($token)) {
            Log::warning('GeoLicense: verification skipped — no token in cache.');
            $this->holder->setValid(false);
            return;
        }

        $serverUrl = config('geolicense.server_url');

        try {
            $response = Http::timeout(15)
                ->acceptJson()
                ->post("{$serverUrl}/api/v1/licenses/verify", [
                    'token'     => $token,
                    'machineId' => $this->machineId->generate(),
                ]);

            $valid = $response->successful() && $response->json('status');
            $this->holder->setValid($valid);

            if (!$valid) {
                $message = $response->json('messages.0', 'unknown reason');
                Log::warning("GeoLicense: verification failed — {$message}");
            } else {
                Log::debug('GeoLicense: license verified successfully.');
            }
        } catch (\Throwable $e) {
            // Do not flip valid to false — let the grace period handle transient outages.
            Log::warning('GeoLicense: server unreachable during verification.', [
                'error' => $e->getMessage(),
            ]);
        }
    }
}
